modules/broker-core/src/main/java/io/ballerina/messaging/broker/core/BindingsRegistry.java

```diff
@@ -43,38 +51,58 @@
 
     private final BindingDeleteListener bindingDeleteListener;
 
+    private List<BindingsRegistryListener> bindingsRegistryListeners;
+
+    private final ReadWriteLock lock;
+
     BindingsRegistry(Exchange exchange, BindingDao bindingDao) {
         this.bindingPatternToBindingsMap = new HashMap<>();
         this.exchange = exchange;
         this.bindingDao = bindingDao;
         this.unmodifiableBindingSetView = Collections.unmodifiableMap(bindingPatternToBindingsMap);
         bindingDeleteListener = new BindingDeleteListener();
+        bindingsRegistryListeners = new ArrayList<>();
+        lock = new ReentrantReadWriteLock();
     }
 
     void bind(QueueHandler queueHandler, String bindingKey, FieldTable arguments) throws BrokerException,
-                                                                                         ValidationException {
-        BindingSet bindingSet = bindingPatternToBindingsMap.computeIfAbsent(bindingKey, k -> new BindingSet());
-        Queue queue = queueHandler.getQueue();
-        Binding binding = new Binding(queue, bindingKey, arguments);
-        boolean success = bindingSet.add(binding);
-
-        if (success) {
-            queueHandler.addBinding(binding, bindingDeleteListener);
-            if (queue.isDurable()) {
-                bindingDao.persist(exchange.getName(), binding);
+            ValidationException {
+        lock.writeLock().lock();
+        try {
+            BindingSet bindingSet = bindingPatternToBindingsMap.computeIfAbsent(bindingKey, k -> new BindingSet());
+            Queue queue = queueHandler.getQueue();
+            Binding binding = new Binding(queue, bindingKey, arguments);
+            boolean success = bindingSet.add(binding);
+
+            if (success) {
+                queueHandler.addBinding(binding, bindingDeleteListener);
+                if (queue.isDurable()) {
+                    bindingDao.persist(exchange.getName(), binding);
+                }
             }
+            LOGGER.debug("Binding added for queue {} with pattern {}", queueHandler, bindingKey);
+            notifyOnBind(bindingKey);
+        } finally {
+            lock.writeLock().unlock();
         }
     }
 
     void unbind(Queue queue, String routingKey) throws BrokerException {
-        BindingSet bindingSet = bindingPatternToBindingsMap.get(routingKey);
-        if (queue.isDurable()) {
-            bindingDao.delete(queue.getName(), routingKey, exchange.getName());
-        }
-        bindingSet.remove(queue);
+        lock.writeLock().lock();
```

Comment:
This also does not need a lock since BrokerImpl holds a lock before propagating the call here.

Reply:
Fixed in cd6b801